src/System.Management.Automation/engine/Subsystem/Prediction/CommandPrediction.cs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Management.Automation.Subsystem
+{
+    /// <summary>
+    /// The class represents the prediction result from a predictor.
+    /// </summary>
+    public sealed class PredictionResult
+    {
+        /// <summary>
+        /// Gets the Id of the predictor.
+        /// </summary>
+        public Guid Id { get; }
+
+        /// <summary>
+        /// Gets the suggestions.
+        /// </summary>
+        public List<string> Suggestions { get; }
+
+        internal PredictionResult(Guid id, List<string> suggestions)
+        {
+            Id = id;
+            Suggestions = suggestions;
+        }
+    }
+
+    /// <summary>
+    /// Provides a set of possible predictions for given input.
+    /// </summary>
+    public static class CommandPrediction
+    {
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the default timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens)
+        {
+            return PredictInput(ast, astTokens, millisecondsTimeout: 20);
+        }
+
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the specified timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <param name="millisecondsTimeout">The milliseconds to timeout.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static async Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens, int millisecondsTimeout)
+        {
+            var predictors = SubsystemManager.GetSubsystems<IPredictor>();
+            if (predictors.Count == 0)
+            {
+                return null;
+            }
+
+            var context = new PredictionContext(ast, astTokens);
+            var cancellationSource = new CancellationTokenSource();
```

Comment:
Consider:

```suggestion
            using var cancellationSource = new CancellationTokenSource();
```

and removing the `try`/`finally`.